src/LocalField/Ring.jl

```diff
@@ -6,117 +6,151 @@
-end
+################################################################################
+#
+#  Parent type etc
+#
+################################################################################
 
-function MaximalOrder(Q::QadicField)
-  return LocalFieldValuationRing{QadicField, QadicFieldElem}(Q)
-end
+parent_type(::Type{LocalFieldValuationRingElem{S, T}}) where {S, T} = LocalFieldValuationRing{S, T}
+elem_type(::Type{LocalFieldValuationRing{S, T}}) where {S, T} = LocalFieldValuationRingElem{S, T}
+is_domain_type(::Type{<: LocalFieldValuationRing}) = true
+is_exact_type(::Type{<: LocalFieldValuationRing}) = false
+
+################################################################################
+#
+#  Field access
+#
+################################################################################
 
-function MaximalOrder(Q::PadicField)
-  return LocalFieldValuationRing{PadicField, PadicFieldElem}(Q)
+_field(R::LocalFieldValuationRing) = R.Q
+
+base_ring(R::LocalFieldValuationRing) = Union{}
+base_ring_type(::Type{<: LocalFieldValuationRing}) = typeof(Union{})
+
+parent(a::LocalFieldValuationRingElem) = a.P
+data(a::LocalFieldValuationRingElem) = a.x
+
+################################################################################
+#
+#  Basic functionality
+#
+################################################################################
+
+characteristic(R::LocalFieldValuationRing) = 0
+
+function zero(Q::LocalFieldValuationRing; precision::Int=precision(Q))
+  return LocalFieldValuationRingElem(Q, zero(_field(Q), precision = precision))
 end
-#integers(Q::QadicField) = ring_of_integers(Q)
-function MaximalOrder(Q::LocalField{S, T}) where {S, T <: Union{EisensteinLocalField, UnramifiedLocalField}}
-  return LocalFieldValuationRing{LocalField{S, T}, LocalFieldElem{S, T}}(Q)
+
+function one(Q::LocalFieldValuationRing; precision::Int=precision(Q))
+  return LocalFieldValuationRingElem(Q, one(_field(Q), precision = precision))
 end
-#integers(Q::PadicField) = ring_of_integers(Q)
 
-valuation_ring(K::NonArchLocalField) = MaximalOrder(K)
+is_zero(a::LocalFieldValuationRingElem) = is_zero(data(a))
+is_one(a::LocalFieldValuationRingElem) = is_one(data(a))
 
-uniformizer(R::LocalFieldValuationRing) = R(uniformizer(R.Q))
+valuation(a::LocalFieldValuationRingElem) = valuation(data(a))
+uniformizer(R::LocalFieldValuationRing) = R(uniformizer(_field(R)))
 
-function Base.show(io::IO, a::LocalFieldValuationRingElem)
-  print(io, a.x)
-end
+is_unit(a::LocalFieldValuationRingElem) = !iszero(a) && valuation(a) == 0
 
-function AbstractAlgebra.expressify(a::LocalFieldValuationRingElem; context=nothing)
-  return expressify(a.x; context=context)
+function canonical_unit(a::LocalFieldValuationRingElem)
+  iszero(data(a)) && return setprecision(parent(a)(1), precision(a))
+  v = ZZ(absolute_ramification_index(_field(parent(a)))*valuation(data(a)))
+  pi = uniformizer(_field(parent(a)))
+  return LocalFieldValuationRingElem(parent(a), data(a)//pi^v)
```

Comment:
The biggest non-trivial change is in `canonical_unit`.
* Before this would return the inverse of what it does now? So an element `u` such that `a*u` is "nicer". The way I understand the AbstractAlgebra documentation, it should be `a//u`.
* The function used the prime instead of the uniformizer, so this wouldn't work for `LocalField` where the valuation might be a fraction.